paddle/framework/channel_test.cc

```diff
@@ -60,13 +60,83 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
   delete ch;
 }
 
-TEST(Channel, SendOnClosedChannelPanics) {
+// This tests that a buffered channel must panic (return false)
+// on send and receive performed after closing the channel.
+// Receive will only panic after close when queue is empty
+TEST(Channel, ClosedBufferedChannelPanics) {
   const size_t buffer_size = 10;
+  const size_t data = 5;
   auto ch = MakeChannel<size_t>(buffer_size);
-  size_t i = 5;
-  EXPECT_EQ(ch->Send(&i), true);  // should not block or panic
+
+  std::thread send_thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), true);  // should not block or panic
+  }};
+
+  std::thread recv_thread{[&]() {
+    size_t i;
+    EXPECT_EQ(ch->Receive(&i), true);  // should not block or panic
+    EXPECT_EQ(i, data);
+  }};
+
+  send_thread.join();
+  recv_thread.join();
+
+  // After closing send should panic. Receive should
+  // also panic because the queue is empty.
   CloseChannel(ch);
-  EXPECT_EQ(ch->Send(&i), false);  // should panic
+
+  send_thread = std::thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), false);  // should panic because channel is closed
+  }};
+  recv_thread = std::thread{[&]() {
+    size_t i;
+    // should panic because channel is closed and queue is empty
+    EXPECT_EQ(ch->Receive(&i), false);
+  }};
+
+  send_thread.join();
+  recv_thread.join();
+  delete ch;
+}
+
+// This tests that an unbuffered channel must panic (return false)
+// on send and receive performed after closing the channel
+TEST(Channel, ClosedUnBufferedChannelPanics) {
+  const size_t data = 5;
+  auto ch = MakeChannel<size_t>(0);
+
+  std::thread send_thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), true);  // should not block or panic
+  }};
+
+  std::thread recv_thread{[&]() {
+    size_t i;
+    EXPECT_EQ(ch->Receive(&i), true);  // should not block or panic
+    EXPECT_EQ(i, data);
+  }};
+
+  send_thread.join();
+  recv_thread.join();
+
+  // After closing send should panic. Receive should
+  // also panic because the queue is empty.
+  CloseChannel(ch);
+
+  send_thread = std::thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), false);  // should panic because channel is closed
+  }};
+  recv_thread = std::thread{[&]() {
+    size_t i;
+    // should panic because channel is closed and queue is empty
+    EXPECT_EQ(ch->Receive(&i), false);
+  }};
+
+  send_thread.join();
+  recv_thread.join();
```

Comment:
Also, I think the function `ClosedUnBufferedChannelPanics` and the function `ClosedBufferedChannelPanics` reuse a block of code as a part of the functionality.
I propose that we write the reused block inside another function and call that function with a buffered and unbuffered channel , check PR: https://github.com/PaddlePaddle/Paddle/pull/8162 for reference.

Reply:
I agreed with @kavyasrinet .

Reply:
Fixed in the latest commit. Thank you